clodius/cli/aggregate.py

```diff
@@ -754,27 +841,159 @@ def line_to_np_array(line):
+    conn,
+    c,
+    sorted_intervals,
+    uid_to_interval,
+    max_zoom,
+    tile_size,
+    tile_counts,
+    max_per_tile,
+    sqlite_cache_size=500,  # 500 MB
+    sqlite_batch_size=100000,
+    verbose=False,
+):
+    if verbose:
+        print("Indexing strategy: tile-based")
+
+    row = c.execute("SELECT * from tileset_info").fetchone()
+    version = row[next(zip(*c.description)).index("version")]
+    c.execute(
+        f"""
+        UPDATE tileset_info
+        SET version = '{version}t'
+        WHERE version = '{version}'
+        """
+    )
     conn.commit()
-    return True
+
+    c.execute(
+        """
+        CREATE TABLE tiles
+        (
+            id int,
+            intervalId int,
+            PRIMARY KEY (id, intervalId)
+        )
+        """
+    )
+
+    # I.e., tiles_cumsum[3] is the number of tiles with zoomlevels lower than 3
+    tiles_cumsum = np.cumsum([0] + [2 ** x for x in range(max_zoom + 1)])
+
+    interval_inserts = []
+    tile_inserts = []
+
+    def batch_insert(conn, c, interval_inserts, tile_inserts, counter):
+        if verbose > 0:
+            print(f"Insert batch ({counter})")
+
+        with transaction(conn):
+            c.executemany(
+                "INSERT INTO intervals VALUES (?,?,?,?,?,?,?,?,?)", interval_inserts
+            )
+            c.executemany("INSERT INTO tiles VALUES (?,?)", tile_inserts)
+
+        interval_inserts.clear()
+        tile_inserts.clear()
+
+    for interval_idx, interval in enumerate(sorted_intervals):
+        curr_zoom = 0
+        inserted = False
+        # go through each interval from most important to least
+        while curr_zoom <= max_zoom:
+            # try to place it in the highest zoom level and go down from there
+            tile_width = tile_size * 2 ** (max_zoom - curr_zoom)
+
+            curr_pos = interval[0]
+            space_available = True
+
+            # if we have not yet inserted the interval somewhere
+            if not inserted:
+                # check if there's space at this zoom level
+                while curr_pos < interval[1]:
+                    curr_tile = math.floor(curr_pos / tile_width)
+                    tile_id = f"{curr_zoom}.{curr_tile}"
+
+                    # if any of the overlapping tiles is already filled up,
+                    # lets go to the next zoom level by setting `space_available`
+                    # to false
+                    if tile_counts[tile_id] >= max_per_tile:
+                        space_available = False
+                        break
+
+                    curr_pos += tile_width
+
+            # if there is, then increment the tile counters
+            if not inserted and space_available:
+                curr_pos = interval[0]
+                while curr_pos < interval[1]:
+                    curr_tile = math.floor(curr_pos / tile_width)
+                    tile_id = "{}.{}".format(curr_zoom, curr_tile)
+
+                    tile_counts[tile_id] += 1
+
```

Comment:
Couldn't you put the `tile_inserts.append()` call from below here and save yourself a while loop?

Reply:
I don't think this works. The reason why is the following:

`tile_counts` only counts the _first_ time an interval is inserted but not the subsequent times. Hence, this `while` loop is only called once, while the one below is called in higher zoom levels as well.

Here's a short example: say we set the max number of intervals per tile to 2. The first interval will be inserted at tile 0.0 and its counter will increase. However, we want to make sure that the most important tile is always shown in tile with a higher zoom level so we will insert tile-id<>interval-id pairs for higher zoom levels as well. However, we will **not** increase the counter for those tiles as otherwise we would never see more details as we zoom in in some cases. Does this somehow make sense? :)

This replicates the existing behavior of the beddb format.